src/engine-actions/translate-swap.js

```diff
@@ -38,40 +50,139 @@ async function translateSwap (takerAddress, swapHash, amount, maxTime, finalCltv
   this.logger.info(`Translating swap for ${swapHash} by sending to ${takerAddress}`,
     { amount, maxTime })
 
-  const { publicKey } = networkAddressFormatter.parse(takerAddress)
-
-  // Convert the maximum time into a time from now so we can convert it into
-  // blocks of relative time. We round down to a whole number of seconds to
-  // be conservative.
-  const timeFromNow = Math.floor((maxTime - new Date()) / 1000)
-
-  // Translate seconds into blocks, rounding down to ensure we can stay atomic.
-  // The LND api expects this to be an int32, which in javascript is a number.
-  const cltvLimit = parseInt(Big(timeFromNow).div(this.secondsPerBlock).round(0, 0).toFixed(0), 10)
-
-  // Translate seconds into blocks, rounding up so that we're sure the end
-  // node will accept the payment. The LND api expects this to be an int32,
-  // so we convert to a javascript number.
-  const finalCltvDelta = parseInt(Big(finalCltvDeltaSecs).div(this.secondsPerBlock).round(0, 3).toFixed(0), 10)
-
-  const {
-    paymentPreimage,
-    paymentError
-  } = await sendPayment({
-    paymentHash: swapHash,
-    destString: publicKey,
-    amt: amount,
-    finalCltvDelta,
-    cltvLimit,
-    feeLimit: DEFAULT_FEE_LIMIT
-  }, { client: this.client })
-
-  if (paymentError) {
-    this.logger.error(`Payment error while sending to route`, { paymentError, swapHash })
-    throw new PermanentSwapError(paymentError)
+  // To ensure translateSwap is idempotent, we first make a call to
+  // getPreimageIfPaymentExists to make sure we don't have an existing payment
+  // that is either In-Flight, Completed, or Failed. If there is not an existing
+  // preimage for the provided swapHash, we can be confident that a call to
+  // sendPayment will result in creating only one downstream payment for this swap
+  try {
+    // we await so rejections are handled within this function
+    const paymentPreimage = await getPreimageIfPaymentExists(swapHash)
+
+    return paymentPreimage
+  } catch (e) {
+    // If our call to LND returned a gRPC NOT FOUND error, we can be confident there
+    // isn't an existing payment for the provided swapHash, so we can create a new
+    // payment with a call to sendPayment
+    if (e.code !== grpc.status.NOT_FOUND) {
+      throw new PermanentSwapError(e)
```

Comment:
How did you decide to make this a permanent swap error? It feels like we're in a bit of an unknown state if this call failed (without a NOT FOUND).

Reply:
Especially since you explicitly mark errors as permanent or not in `getPreimageIfPaymentExists` if feels odd to re-categorize it here.